misc/observerability-symbol-collector/src/main/java/org/ballerinalang/observability/anaylze/DefaultObservabilitySymbolCollector.java

```diff
@@ -0,0 +1,163 @@
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Default implementation of ObserverbilitySymbolCollector.
+ */
+public class DefaultObservabilitySymbolCollector implements ObservabilitySymbolCollector {
+    private final PrintStream console = System.out;
+
+    private static final String NAME = "name";
+    private static final String ORG_NAME = "orgName";
+    private static final String PKG_VERSION = "pkgVersion";
+    private static final String COMPILATION_UNITS = "compilationUnits";
+    private static final String SRC = "src";
+    private static final String KEY_URI = "uri";
+    private static final String AST = "ast";
+    private static final String JSON = ".json";
+    public static final String PROGRAM_HASH_KEY = "PROGRAM_HASH";
+    public static final String AST_META_FILENAME = "meta.properties";
+
+    private CompilerContext compilerContext;
+
+    private DiagnosticLog diagnosticLog;
+
+    @Override
+    public void init(CompilerContext context) {
+        compilerContext = context;
+        diagnosticLog = BLangDiagnosticLogHelper.getInstance(context);
+    }
+
+    @Override
+    public void process(BLangPackage module) {
+        JsonObject moduleJson = getModuleJson(module);
+        JsonASTHolder.getInstance().addAST(module.packageID.name.getValue(), moduleJson);
+    }
+
+    @Override
+    public void writeCollectedSymbols(BLangPackage module, Path destination) throws IOException {
+        Path targetDirPath = destination.resolve(AST);
+        if (Files.notExists(targetDirPath)) {
+            Files.createDirectory(targetDirPath);
+        }
+        Set<String> packages = getUserPackages();
+
+        JsonObject ast = new JsonObject();
+        for (Map.Entry<String, JsonObject> entry : JsonASTHolder.getInstance()
+                .getASTMap().entrySet()) {
+            if (packages.contains(entry.getKey())) {
+                ast.add(entry.getKey(), entry.getValue());
+            }
+        }
+
+        String astDataString = JsonCanonicalizer.getEncodedString(ast.toString());
+        Files.write(targetDirPath.resolve(AST + JSON), astDataString.getBytes(StandardCharsets.UTF_8));
+
+        Properties props = new Properties();
+        props.setProperty(PROGRAM_HASH_KEY, String.valueOf(astDataString.hashCode()));
+        try (OutputStream outputStream = Files.newOutputStream(targetDirPath.resolve(AST_META_FILENAME))) {
+            props.store(outputStream, null);
+        }
+    }
+
+    private Set<String> getUserPackages() {
+        SourceDirectoryManager sourceDirectoryManager = SourceDirectoryManager.getInstance(compilerContext);
+        return sourceDirectoryManager.listSourceFilesAndPackages()
```

Comment:
We do not recommend to use stream api if performance is critical

Reply:
This is not in the critical path AFAIU. Is it ok if I keep this for readability?